### ISSUE TITLE
Report error spans to error listeners

### ISSUE THREAD

**Issue by michaelpj** [antlr/antlr4#940]
(Previously discussed in https://github.com/antlr/antlr4/pull/888)

It would be nice to give both start and end line/index pairs when reporting errors to listeners. These are useful for highlighting spans in editors, and a little tricky to work out from the recognizer directly.

---

Let me give a recent example of such trickiness. I have been attempting to get end positions for lexer errors from the `Lexer` itself, as suggested. However, it appears that `Lexer.getCharPositionInLine()` behaves differently for single unrecognized characters than it does for characters which form part of an incomplete token.

For example, consider the following lexer grammar fragment, which matches typical `"`-delimited string literals:
```
String : '"' (~[\\\"\n\r\t] | '\\' [\"nrt\\] )* '"';
```
On the input `"`, the listener is triggered with `charPositionInLine = 0`, and `Lexer.getCharPositionInLine() = 1`. However, on the input `#`, `Lexer.getCharPositionInLine()` is instead `0`.

I doubt this is really a *bug*, since I'm poking around in the internal state of the recognizer, which likely wasn't intended to be used like this. But I hope it supports my case that it's tricky for the user to do this correctly, and it would be better if it were provided by the interface.

**Comment by parrt** [comment]
I'll look at this for the upcoming release. thanks!

**Comment by parrt** [comment]
Actually those character positions make sense to me because `"` has started to match a token whereas `#` is not in any token and so its position is in fact 0.

**Comment by michaelpj** [comment]
The text in question begins at 0 and ends at 1 for both of them, though. Indeed, the error message that is produced correctly reports the erroneous text as `"` and `#` respectively, so we do know which bit of text is the problem.

For the purposes of reporting errors, you want positions *in the text*, regardless of how well tokenization has gone. In fact, you're probably expecting strange lexer/parser state, because you're in a failure mode! I understand why we currently get the positions that we do, I just don't think they're useful for error reporting, and we should try and just report what the user is likely to want: the position span of the offending piece of input.

**Comment by sharwell** [comment]
@michaelpj You are taking a very roundabout approach to reaching my conclusion that your lexer should be able to tokenize all inputs without an error, and just let the parser report the error(s) from the invalid tokens. :wink: It really does work well.

**Comment by sharwell** [comment]
> The text in question begins at 0 and ends at 1 for both of them, though.

In the case of `"`, the character was consumed because it could have been part of a token (even though the token wasn't complete). For `#`, no token can start with that character so the listener was invoked before consuming it. The fact that the `#` token was consumed despite not matching any transition is an intentional behavior to prevent a loop of match &rarr; fail &rarr; retry match which produces an infinite sequence of zero-length tokens (a situation which was actually possible in ANTLR 3).

**Comment by sharwell** [comment]
If your lexer can turn any input text into tokens, there is no question about what your error handler will see. It will always be a sequence of fully-matched tokens produced by your lexer (some of which were likely created to represent erroneous input). If you used the easiest approach of an `ERR_CHAR` rule (which precisely matches the default error recovery behavior of lexers), those two cases would each produce an `ERR_CHAR` token covering the input span you desired.

**Comment by michaelpj** [comment]
I agree that if you have a lexer that cannot fail, then it is very easy to do lexer error messages :wink:

I still think this is an issue for two reasons:

1. We *do* provide lexer errors, and we *do* have an interface for consuming them. If we can make the errors we emit more helpful for the user, why not?
2. It's nice to have better error reporting *without* having to construct your lexer differently!

> In the case of ", the character was consumed because it could have been part of a token (even though the token wasn't complete). For #, no token can start with that character so the listener was invoked before consuming it.

Why should the user care about that difference? In both cases the problem is the same: we saw a character which we were unable to use to form a token. The locations of the characters are the same. From a user's point of view, those errors are the same, apart from the particular character that is in error.

---

Throughout this I have in my head the happy path for an ANTLR user who wants to highlight errors in `$EDITOR` (i.e. me). I imagine it might go like this:

1. Create an `ANTLRErrorListener`.
2. Implement `syntaxError`:
    1. Note down the start line/column and end line/column, use this to create a red squiggly underline in your editor.
    2. Attach the error message to the squiggle.
3. Attach your listener to your `Lexer` and `Parser`.

That's easy; handles errors from `Lexer`s and `Parser`s consistently; doesn't require you to change your `Lexer`; and doesn't require any understanding of the internals.

With that as the goal, I don't really think this is about the details of when we consume tokens - that's an implementation detail. Ideally we'd give good error messages in all circumstances!

---

Finally, here's a suggestion for a method for lexer error spans:

- If we consumed any characters between the last token and getting stuck, produce an span that includes all of them.
- If we were unable to consume a character at all, produce a span covering the next character.

For parsers, we could just use the span of the erroneous token, which is straightforward enough in principle (in practice, I don't believe the end positions of tokens are currently recorded, which makes working out the correct ending line a little tricky).

**Comment by sharwell** [comment]
> We do provide lexer errors, and we do have an interface for consuming them.

This was partly my mistake. We should have made it very clear from the start that the purpose of this mechanism is to aid in grammar debugging by identifying cases where a lexer failed to account for some input.

> It's nice to have better error reporting without having to construct your lexer differently!

It's still going to be a problem, because the lexer has exceptionally poor error recovery for tokens containing more than one character. For example, consider a case where the lexer is used for syntax highlighting and the user types the following incomplete statement:

```java
String foo = "some text
```

In this case, the parser will see the following, where `some` and `text` are treated as identifiers instead of as part of an obviously incomplete string. The syntax highlighting will be completely wrong.

```java
String foo = some text
```

Even if you figure out how to get consistent error reports from the lexer, and even if we make changes to the error reporting structure to support lexers that can't understand certain input, you will still get overwhelmingly better behavior by writing your lexer to support other input.

---

:thought_balloon: One possibility for improving the behavior is creating a new reserved token type for error characters, and creating an implicit token of this type *instead* of reporting lexer errors.

**Comment by jimidle** [comment]
We have thrashed this to death over many years. What Sam is saying is
basically "the way" and for more than just practical reasons. Your focus on
the error message coming directly at lexing point turns out to violate a
useful 'principle' of constructing front ends, which is basically to defer
error messages/handling to the latest point in the sequence that you can
(see emails for me to the group passim ad nauseam).

The reason to do this is because of always having better context, with the
more information content you have created. As an input, the 'program' is
just a sequence of characters. The lexer adds more information by
tokenizing, the parser adds more information by verifying token order and
constructing a parsing model, traversing that model allows semantic
checking....

So to repeat myself again ;), your lexer should always at minimum end with:

BAD: . ;

Which you will be fed as a syntax error in the parser, and having more
information, you can give your user a better error message. For the lexer,
you just get:

Invalid character '.' at line 42, offset 42

But in your parser, you can, should you feel it helps the user, say
something like:

Error(42,42) at '.': It looks like you have a '.' in your comma separated
list - typo?

And of course more realistic scenarios abound when you look at the language
task at hand.

For similar reasons, you parser should accept pretty much anything that
COULD be valid input:

seq: s+=ID (',' s+=ID);

instead of

seq: ID ',' ID;

As, assuming that this does not affect parsing of valid constructs, you can
now use this rule for any number of arguments then SEMANTICALLY issue
better message. So instead of:

Error(42, 84) at ',' - Syntax error, expecting ';'

You can now say:

Error(42, 84) at '.': Sequences for zarquon statements should have two
elements, but here you have supplied 42

In all cases, the higher up the chain you can go before reporting the
error, then the more information that you have (or at least the more
readily available is that information) and the better errors you can issue.

Hence, it follows that whenever possible, lexers should not issue errors at
all. But there are some instances where it makes sense:

STRING      : '"' (~( '\n' | '\r' | '"' ))*
                (   ( '\n' | '\r' ) { /* TODO: Implement unterminated
literal error */ }
                  | '"'
                )
            ;

However, even that can be managed better by just changing the token that
the lexer supplies to UNTERMINATEDSTRING and having the parser handle this
with a better error message involving context, or even accepting that as
well as STRING (but with an error message) so that it can continue to look
at more source and report as many errors as it can in one pass at the
input.

There is nothing worse than:

$ compile myblob.grr
error at line 7, ':' unexpected.
$ vi ... # fix line 7
$ compile myblob.grr
error at line 8, ".' unexpected.
$ vi .... # grrrrr
.... repeat

vs:
$ compile myblob.grr
Error(7, 8): near ':': Were you trying to make a namespace reference
Jim::Obj ?
Error(8, 12): near '.' : Variable names cannot contain ','
$ vi ... # fix all errors and it's obvious what they are

So, in the end this is what Sam is getting at, though he is using the
example of syntax highlighting. With syntax highlighting, you are expecting
that more often than not, the input will appear to be erroneous because it
is not yet complete. So you want to push the errors up the line until you
have a point where there is context that helps you provide hints and stuff.


Hope that helps :)

Jim

**Comment by jimidle** [comment]
On Thu, Jul 2, 2015 at 10:39 AM, Jim Idle <jimi@temporal-wave.com> wrote:

>
>
> For similar reasons, you parser should accept pretty much anything that
> COULD be valid input:
>
> seq: s+=ID (',' s+=ID);
>


That should have been:

seq: s+=ID (',' s+=ID)*;

**Comment by michaelpj** [comment]
I agree that you get better errors the later you report them. I plan to take up several of your suggestions (indeed, I already rewrote my lexer to be much more permissive :wink:).

I don't see this as inconsistent with arguing that the errors we do give should be as helpful as possible (and note that I'm arguing for the same interface for parser errors too). Users should defer error handling from the lexer if they want more semantically sensical errors, not because they're struggling with the positions! I would like the lexer errors to be better than they are, not necessarily excellent.

Even if you just view lexer errors as a debugging aid, having better positional information is nice to have.
